sw/host/opentitanlib/src/bootstrap/rescue.rs

```diff
@@ -0,0 +1,352 @@
+            .unwrap_or(0)
+            + 1)
+            * 4;
+
+        let mut frames = Vec::new();
+        let mut frame_num = 0;
+        let mut addr = min_addr;
+        while addr < max_addr {
+            // Try skipping over 0xffffffff words.
+            let nonempty_addr = addr
+                + payload[addr..]
+                    .chunks(4)
+                    .position(|c| c != &[0xff; 4])
+                    .unwrap()
+                    * 4;
+            let skip_addr = nonempty_addr & !Self::FLASH_SECTOR_MASK;
+            if skip_addr > addr && (addr == 0 || addr & Self::FLASH_BUFFER_MASK != 0) {
+                // Can only skip from the start or if the last addr wasn't an exact multiple of
+                // 128 (per H1D boot rom).
+                addr = skip_addr;
+            }
+
+            let mut frame = Frame {
+                header: FrameHeader {
+                    frame_num,
+                    flash_offset: addr as u32,
+                    ..Default::default()
+                },
+                ..Default::default()
+            };
+            let slice_size = Self::DATA_LEN.min(payload.len() - addr);
+            frame.data[..slice_size].copy_from_slice(&payload[addr..addr + slice_size]);
+            frames.push(frame);
+
+            addr += Self::DATA_LEN;
+            frame_num += 1;
+        }
+        frames.last_mut().map(|f| f.header.frame_num |= Self::EOF);
+        frames
+            .iter_mut()
+            .for_each(|f| f.header.hash = f.header_hash());
+        Ok(frames)
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum RescueError {
+    #[error("Unrecognized image file format")]
+    ImageFormatError,
+    #[error("Synchronization error communicating with boot rom")]
+    SyncError,
+    #[error("Repeated errors communicating with boot rom")]
+    RepeatedErrors,
+}
+
+/// Implements the UART rescue protocol of Google Ti50 firmware.
+pub struct Rescue {}
+
+impl Rescue {
+    const MAX_CONSECUTIVE_ERRORS: u32 = 50;
+
+    /// Creates a new `Rescue` protocol updater from `options`.
+    pub fn new(_options: &BootstrapOptions) -> Self {
+        Self {}
+    }
+
+    fn read_char(&self, uart: &dyn Uart) -> Result<char> {
+        let mut buf = [0u8; 1];
+        match uart.read_timeout(&mut buf, Duration::from_millis(100)) {
+            Ok(1) => Ok(buf[0] as char),
+            Ok(_) => Err(anyhow!("")),
+            _ => Err(anyhow!("")),
+        }
+    }
+
+    fn expect_string(&self, uart: &dyn Uart, s: &str) -> Result<()> {
+        let char_vec: Vec<char> = s.chars().collect();
+        let mut char_vec: &[char] = &char_vec;
+        while char_vec.len() > 0 {
+            if self.read_char(uart)? != char_vec[0] {
+                return Err(anyhow!(""));
+            }
+            char_vec = &char_vec[1..];
+        }
+        Ok(())
+    }
+
+    fn flush_rx(&self, uart: &dyn Uart) {
+        let mut response = [0u8; Frame::HASH_LEN];
+        loop {
```

Comment:
I think a more idiomatic way to write this might be...
```
while let Ok(len) = uart.read_timeout() {
  if len == 0 { break }
}
```

Reply:
Your suggestion does not consider the possibility of Err(...) return from read_timeout, which I want to treat the same as Ok(0) as the likely error is timeout.  (Some implementation return Ok(0) on timeout, some return some Err(...).)